### DISCUSSION TITLE
Testing authentication

### DISCUSSION THREAD

**Discussion by mktcode** [rails/rails#53905] [Docs]
The [Rails testing documentation](https://guides.rubyonrails.org/testing.html) is excellent, but I find it lacks sufficient information on testing authentication. After following the [introduction video](https://www.youtube.com/watch?v=X_Hw9P1iZfQ), I ran `rails g authentication`, but my tests—both controllers and system tests—started failing. This is because the pages now require users to be logged in, but the video doesn’t cover testing at all.

For acceptance tests, I’d prefer not to go through the login process in every test. While I want to test the login functionality in a few cases, repeatedly filling in the login form in all tests is inefficient. I explored resources like [this guide for testing with Capybara](https://github.com/heartcombo/devise/wiki/How-To:-Test-with-Capybara) and [this Warden testing guide](https://github.com/wardencommunity/warden/wiki/Testing), but I couldn’t get them to work. I’m unsure about what exactly needs to be included or required and where, or whether Ruby already provides everything I need, or if I need to install anything.

As someone new to Rails and Ruby in general, I’m still getting familiar with the ecosystem. A dedicated section in the Rails documentation on testing authentication, including examples for both system and controller tests, would be incredibly helpful.

---

My solution right now is this:

```rb
# test/helpers/login.rb
def login_as(user)
  fill_in "Email", with: user.email_address
  fill_in "Password", with: "password"
  click_button "Sign in"
end
```

```rb
# test/application_system_test_case.rb
require "test_helper"
require_relative "helpers/trix"
require_relative "helpers/login"

class ApplicationSystemTestCase < ActionDispatch::SystemTestCase
  driven_by :selenium, using: :headless_chrome, screen_size: [ 1400, 1400 ]
end
```

```rb
require "application_system_test_case"

class ProductsTest < ApplicationSystemTestCase
  setup do
    @product = products(:one)
    @user = users(:one)
  end

  test "visiting the index" do
    visit products_url
    login_as @user
    assert_selector "h1", text: "Products"
  end
end
```

And for my controllers:

```rb
require "test_helper"

class ProductsControllerTest < ActionDispatch::IntegrationTest
  setup do
    @product = products(:one)
    @user = users(:one)
  end

  test "should get index" do
    post session_url, params: { email_address: @user.email_address, password: "password" }
    get products_url
    assert_response :success
  end

  test "should get new" do
    post session_url, params: { email_address: @user.email_address, password: "password" }
    get new_product_url
    assert_response :success
  end
  
  # ...

end
```

**Comment by dwightwatson** [comment]
An alternative approach is a test helper that replicates the cookie used by the `Authentication` concern.

```rb
def authenticate_as(user)
  cookie_jar = ActionDispatch::Request.new(Rails.application.env_config.deep_dup).cookie_jar
  user.sessions.create!.tap do |session|
    cookie_jar.signed.permanent[:session_id] = { value: session.id, httponly: true, samesite: :lax }
    cookies[:session_id] = cookie_jar[:session_id]
  end
end
```
